packages/core/contracts/financial-templates/optimistic-rewarder/OptimisticRewarder.sol

```diff
@@ -0,0 +1,429 @@
+            "ancillary data too long"
+        );
+        customAncillaryData = _customAncillaryData;
+    }
+
+    /**
+     * @notice Allows anyone to deposit reward tokens into the contract. Presumably, this would be the deployer or
+     * protocol that wishes to reward the users interacting with the system.
+     * @dev Once tokens are deposited, they cannot be withdrawn without claiming a reward. If a deployer wants an
+     * "escape hatch", they can create a special tokenId for this purpose.
+     * @param token ERC20 token that is being deposited.
+     * @param amount amount of rewards to deposit.
+     */
+    function depositRewards(IERC20 token, uint256 amount) public nonReentrant {
+        token.safeTransferFrom(msg.sender, address(this), amount);
+        emit Deposited(msg.sender, token, amount);
+    }
+
+    /**
+     * @notice Allows the caller to mint a token to the receiver and include a reward-relevant update event with it.
+     * This is intended to be used when the user first interacts with a reward-granting protocol.
+     * @dev if the user prefers to only mint a new token, they should call the mintNextToken function.
+     * @param receiver user that will receive the newly minted token.
+     * @param data arbitrary caller-generated data that will be associated with this update.
+     */
+    function mint(address receiver, bytes memory data) public nonReentrant returns (uint256 tokenId) {
+        tokenId = mintNextToken(receiver);
+        emit UpdateToken(tokenId, msg.sender, data);
+    }
+
+    /**
+     * @notice Applies a reward-relevant update to an existing token.
+     * @param tokenId the existing tokenId that the update should be applied to.
+     * @param data arbitrary caller-generated data that will be associated with this update.
+     */
+    function updateToken(uint256 tokenId, bytes memory data) public nonReentrant {
+        emit UpdateToken(tokenId, msg.sender, data);
+    }
+
+    /**
+     * @notice Requests a redemption for any tokenId. This can be called by anyone.
+     * @dev if called by someone who doesn't own the token, they are effectively gifting their bond to the owner.
+     * @param tokenId the tokenId the redemption is for.
+     * @param cumulativeRedemptions the cumulative token addresses and amounts that this tokenId is eligible for
+     * at the current timestap. cumulative redemptions that are too low should be considered to be valid.
+     */
+    function submitRedemption(uint256 tokenId, RedemptionAmount[] memory cumulativeRedemptions) public nonReentrant {
+        bytes32 redemptionId = _redemptionId(tokenId, cumulativeRedemptions);
+        require(redemptions[redemptionId].expiryTime == 0, "Redemption already exists");
+        require(ownerOf(tokenId) != address(0), "tokenId is invalid");
+
+        uint256 time = getCurrentTime();
+
+        uint256 finalFee = _getStore().computeFinalFee(address(bondToken)).rawValue;
```

Comment:
would it be useful to store the final fee in this contract so we dont need to make external calls to the `store` on every `request`? this would be similar to how the BridgePool stores fee information.

Reply:
Can use a similar `sync` method

Reply:
Done.